provider/pkg/helm/tool.go

```diff
@@ -0,0 +1,516 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// package helm contains code vendored from the upstream Helm project.
+package helm
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi-kubernetes/provider/v4/pkg/logging"
+	helmv4 "github.com/pulumi/pulumi-kubernetes/sdk/v4/go/kubernetes/helm/v4"
+	logger "github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/chartutil"
+	"helm.sh/helm/v3/pkg/cli"
+	"helm.sh/helm/v3/pkg/downloader"
+	"helm.sh/helm/v3/pkg/getter"
+	"helm.sh/helm/v3/pkg/registry"
+	"helm.sh/helm/v3/pkg/release"
+	"k8s.io/client-go/util/homedir"
+)
+
+type InitActionConfigF func(actionConfig *action.Configuration, namespaceOverride string) error
+
+type LocateChartF func(i *action.Install, name string, settings *cli.EnvSettings) (string, error)
+
+type ExecuteF func(ctx context.Context, i *action.Install, chrt *chart.Chart, vals map[string]interface{}) (*release.Release, error)
+
+// Tool for executing Helm commands via the Helm library.
+type Tool struct {
+	EnvSettings *cli.EnvSettings
+	HelmDriver  string
+
+	initActionConfig InitActionConfigF
+	locateChart      LocateChartF
+	execute          ExecuteF
+}
+
+// NewTool creates a new Helm tool with the given environment settings.
+func NewTool(settings *cli.EnvSettings) *Tool {
+	helmDriver := os.Getenv("HELM_DRIVER")
+	logger.V(6).Infof("initializing Helm tool: driver=%q, settings=%+v", helmDriver, *settings)
+
+	return &Tool{
+		EnvSettings: settings,
+		HelmDriver:  helmDriver,
+		initActionConfig: func(actionConfig *action.Configuration, namespaceOverride string) error {
+			if namespaceOverride == "" {
+				namespaceOverride = settings.Namespace()
+			}
+			// https://github.com/helm/helm/blob/635b8cf33d25a86131635c32f35b2a76256e40cb/cmd/helm/helm.go#L72-L81
```

Comment:
I sort of agree that it may be possible, but it might might be some package globals to contend with, and I am a bit worried about being "too" arms-length in case we need some Pulumi specifics in the future.  I think the current approach strikes a good balance, maybe we can abstract further when we do Release v4.

Reply:
I feel ya, the Docker provider is in a very similar boat!